trio/_subprocess/unix_pipes.py

```diff
@@ -0,0 +1,107 @@
+import os
+from typing import Tuple
+
+from .. import _core, BrokenStreamError
+from .._abc import SendStream, ReceiveStream
+
+__all__ = ["PipeSendStream", "PipeReceiveStream", "make_pipe"]
+
+
+class _PipeMixin:
+    def __init__(self, pipefd: int):
+        if not isinstance(pipefd, int):
+            raise TypeError(
+                "{0.__class__.__name__} needs a pipe fd".format(self)
+            )
+
+        self._pipe = pipefd
+        self._closed = False
+
+        import fcntl
+        flags = fcntl.fcntl(self._pipe, fcntl.F_GETFL)
+        fcntl.fcntl(self._pipe, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+    async def aclose(self):
+        if not self._closed:
+            _core.notify_fd_close(self._pipe)
+            self._closed = True
+            os.close(self._pipe)
+
+        await _core.checkpoint()
+
+    def fileno(self) -> int:
+        """Gets the file descriptor for this pipe."""
+        return self._pipe
+
+
+class PipeSendStream(_PipeMixin, SendStream):
+    """Represents a send stream over an os.pipe object."""
+
+    async def send_all(self, data: bytes):
+        if self._closed:
+            await _core.checkpoint()
+            raise _core.ClosedResourceError("this pipe is already closed")
+
+        if not data:
+            await _core.checkpoint()
+            return
+
+        length = len(data)
+        # adapted from the SocketStream code
+        with memoryview(data) as view:
+            total_sent = 0
+            while total_sent < length:
+                with view[total_sent:] as remaining:
+                    try:
+                        total_sent += os.write(self._pipe, remaining)
+                    except BrokenPipeError as e:
+                        await _core.checkpoint()
+                        raise BrokenStreamError from e
+                    except BlockingIOError:
+                        pass
+
+                await self.wait_send_all_might_not_block()
```

Comment:
I think the way I'd structure this is:

* Just do a `checkpoint` as the very first thing, given how this is structured it doesn't seem worth the hassle of avoiding this.

* Structure the main loop like:

```python
while True:
    <try to write some data>
    <check if we're done; if so, break out of the loop>
    <wait for the pipe to be writable>
```